src/inabox/inabox-viewport.js

```diff
@@ -0,0 +1,205 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {viewerForDoc} from '../viewer';
+import {Viewport, ViewportBindingDef} from '../service/viewport-impl';
+import {getServiceForDoc} from '../service';
+import {resourcesForDoc} from '../../src/resources';
+import {layoutRectLtwh} from '../layout-rect';
+import {Observable} from '../observable';
+import {MessageType} from '../../src/3p-frame';
+import {IframeMessagingClient} from '../../3p/iframe-messaging-client';
+import {dev} from '../log';
+
+/** @const {string} */
+const TAG = 'inabox-viewport';
+
+/**
+ * Implementation of ViewportBindingDef that works inside an non-scrollable
+ * iframe box by listening to host doc for position and resize updates.
+ *
+ * @visibleForTesting
+ * @implements {ViewportBindingDef}
+ */
+export class ViewportBindingInabox {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const {!Window} */
+    this.win = win;
+
+    /** @private @const {!Observable} */
+    this.scrollObservable_ = new Observable();
+
+    /** @private @const {!Observable} */
+    this.resizeObservable_ = new Observable();
+
+    const boxWidth = win./*OK*/innerWidth;
+    const boxHeight = win./*OK*/innerHeight;
+
+    /**
+     * The current viewport rect.
+     * Before hearing from host doc, we're blind about the viewport position
+     * and iframe position. 0 scroll is not a bad guess.
+     * Meanwhile, use iframe box size as the viewport size gives a good
+     * initial resource scheduling.
+     * @private {!../layout-rect.LayoutRectDef}
+     */
+    this.viewportRect_ = layoutRectLtwh(0, 0, boxWidth, boxHeight);
+
+    /**
+     * The current layout rect of the iframe box.
+     * To not trigger amp-analytics visibility immediately,
+     * we start with an initial position right below the fold.
+     * @private {!../layout-rect.LayoutRectDef}
+     */
+    this.boxRect_ = layoutRectLtwh(0, boxHeight + 1, boxWidth, boxHeight);
+
+    /** @private @const {!IframeMessagingClient} */
+    this.iframeClient_ = new IframeMessagingClient(win);
+    this.iframeClient_.setSentinel(getRandom(win));
+
+    // Bet the top window is the scrollable window and loads host script.
+    // TODOs:
+    // 1) check window ancestor origin, if the top window is in same origin,
+    // don't bother to use post messages.
+    // 2) broadcast the request
+    this.iframeClient_.setHostWindow(win.top);
+
+    dev().fine(TAG, 'initialized inabox viewport');
+  }
+
+  /** @override */
+  connect() {
+    this.iframeClient_.makeRequest(
+        MessageType.SEND_POSITIONS, MessageType.POSITION,
+        data => {
+          dev().fine(TAG, 'Position changed: ', data);
+          const oldViewportRect = this.viewportRect_;
+          const oldSelfRect = this.boxRect_;
+          this.viewportRect_ = data.viewport;
+          this.boxRect_ = data.target;
+          if (isMoved(this.boxRect_, oldSelfRect)) {
```

Comment:
Added `isResized` check as well. And added a TODO for firing a DOM mutation event when we handle them.